### PR TITLE
Add : feed icon to rss2.xml

### PR DESCRIPTION
**Pull request by sehajyang**
I configured `feed.icon` in the `config.yml` file.
but that icon was not exposed when creating the feed xml file.
So, I add in `rss2.xml` :
``` 
<url> {{(url + config.feed.icon) | uriencode}} </ url>
```
and I see that it works.

### CODE REVIEW

**Comment by sehajyang** [comment]
`full_url_for ()` seems to add the blog url as a prefix. 
So if it's a third-party url, it seems to be `blog url + third-party url`. am I right?

**Comment by SukkaW** [comment]
@sehajyang Nope.

https://github.com/hexojs/hexo-util/blob/d9081722c3220e7943f073e84ff3cf10db499a8a/lib/full_url_for.js#L21

**Comment by sehajyang** [comment]
@Sukka

I installed hexo-util and added the following code to `generator.js`:
```
const all_url_for = require ('hexo-util'). all_url_for.bind (hexo);

var icon;
if (config.email) {
    icon = gravatar (config.email)
} else {
    icon = all_url_for (config.feed.icon)
}
```
However, there was an error where `config` could not be found in `all_url_for ()`.
Also `bind (hexo);` didn't work. I tried a few ways to solve this problem but it failed.
How can i use `all_url_for`?

**Comment by SukkaW** [comment]
@sehajyang Try this?

First,  require `full_url_for` at the beginning:

https://github.com/hexojs/hexo-generator-feed/blob/d169992a27181e4ab3443bcaaf7b71496113f4c7/lib/generator.js#L7

```js
 const { encodeURL, gravatar, full_url_for } = require('hexo-util'); 
```

and change L38:

https://github.com/hexojs/hexo-generator-feed/blob/d169992a27181e4ab3443bcaaf7b71496113f4c7/lib/generator.js#L38

to: 

```js
if (feedConfig.icon) icon = encodeURI(full_utl_for.call(this, feedConfig.icon)); 
```

You can also include shorter object syntax in this PR:

https://github.com/hexojs/hexo-generator-feed/blob/d169992a27181e4ab3443bcaaf7b71496113f4c7/lib/generator.js#L41-L47

to

```js
 const xml = template.render({ 
   config, 
   url, 
   icon, 
   posts, 
   feed_url: config.root + feedConfig.path 
 }); 
```

**Comment by sehajyang** [comment]
@SukkaW 
Thanks to your comment I was able to solve this issue. Thank you for your help 😊

File: lib/generator.js [hunk 1]
```diff
@@ -35,14 +35,14 @@ module.exports = function(locals) {
   if (url[url.length - 1] !== '/') url += '/';
 
   let icon = '';
-  if (feedConfig.icon) icon = url + encodeURI(feedConfig.icon);
+  if (feedConfig.icon) icon = encodeURI(full_url_for.call(this, feedConfig.icon));
```
Comment (curbengh) [inline lib/generator.js]: `full_url_for()` [already encodes](https://github.com/hexojs/hexo-util/blob/42b7393b9c591d3874e0c687b4dba5a17512121a/lib/full_url_for.js#L24) the url, so `encodeURI()` is not needed.

File: rss2.xml [hunk 2]
```diff
@@ -4,6 +4,7 @@
   xmlns:content="http://purl.org/rss/1.0/modules/content/">
   <channel>
     <title>{{ config.title }}</title>
+    <url>{{ icon }}</url>
```
Comment (curbengh) [inline rss2.xml]: [The spec](http://www.rssboard.org/rss-specification) mentions `<url>` is required to be wrapped in `<image>`, and `<image>` must include `<title>`, `<link>` and `<url>`.

[This tutorial](https://www.xul.fr/en-xml-rss.html) also mentions `<image>`.

``` html
<title>{{ config.title }}</title>
<link>{{ url }}</link>
<image>
  <url>{{ icon }}</url>
  <title>{{ config.title }}</title>
  <link>{{ url }}</link>
</image>
```

**Comment by curbengh** [comment]
Unit test is also missing, let me know if you need help creating one.

**Comment by herrenofficial** [comment]
@curbengh 
Yes please make one.
Also, I modified the part you commented above.  Let me know if there's anything I need to change.

**Comment by sehajyang** [comment]
@curbengh 
Yes please make one.
Also, I modified the part you commented above.  Let me know if there's anything I need to change.

File: rss2.xml [hunk 3]
```diff
@@ -5,6 +5,11 @@
   <channel>
     <title>{{ config.title }}</title>
     <link>{{ url }}</link>
+    <image>
+      <url>{{ icon }}</url>
+      <title>{{ config.title }}</title>
+      <link>{{ url }}</link>
+    </image>
```
Comment (curbengh) [inline rss2.xml]: I just noticed atom.xml does check [existence of `icon`](https://github.com/hexojs/hexo-generator-feed/blob/d169992a27181e4ab3443bcaaf7b71496113f4c7/atom.xml#L4), so rss2 also need.

**Comment by curbengh** [comment]
I just added two tests, for atom only. Simply duplicate them (not replace) and change the necessary config to `rss2`.

A brief guide for using cheerio; atom.xml starts with the following value,

``` xml
<feed xmlns="https://www.w3.org/2005/Atom">
  <title>{{ config.title }}</title>
  <icon>{{ icon }}</icon>
```

Notice `$('feed>icon')` corresponds to the `<feed>` and `<icon>` elements. The `>` is a shortcut for [`.children()`](https://cheerio.js.org/); in this case, we're looking for `<icon>` in the children of `<feed>`.

File: atom.xml [hunk 4]
```diff
@@ -1,7 +1,13 @@
...
 <feed xmlns="https://www.w3.org/2005/Atom">
   <title>{{ config.title }}</title>
-  {% if icon %}<icon>{{ icon }}</icon>{% endif %}
+  {% if icon %}
+  <image>
+    <url>{{ icon }}</url>
+    <title>{{ config.title }}</title>
+    <link>{{ url }}</link>
+  </image>
+  {% endif %}
```
Comment (curbengh) [inline atom.xml]: please revert this change, rss2 spec doesn't apply to atom.

Comment (sehajyang) [reply atom.xml]: Oh.. OK I'll revert that commit

File: rss2.xml (hunk 3 above)
Comment (sehajyang) [reply rss2.xml]: I added part to check if the icon exists (`{% if icon %}`) in rss2.xml

**Comment by sehajyang** [comment]
@curbengh 
I created the rss2 icon test code with your comments and the atom icon test code. 
Please let me know if there are any errors.

Thank you for your help
